src/gradient.rs

```diff
@@ -1,43 +1,39 @@
-use crate::utils::clamp_f32;
 use crate::utils::map_t_of_range_a_to_range_b;
 use crate::utils::Interpolate;
 use itertools::Itertools;
 
-type Rgb = [u8; 3];
-type TColor = (f32, Rgb);
+type Rgba = [u8; 4];
+type TColor = (f32, Rgba);
 
 struct Spectrum(Vec<TColor>);
 
 impl Spectrum {
-    pub fn from_colors(color_a: Rgb, color_b: Rgb) -> Self {
-        let mut spectrum = Vec::new();
-        spectrum.push((0f32, color_a));
-        spectrum.push((1f32, color_b));
-
-        Self(spectrum)
+    pub fn from_colors(color_a: Rgba, color_b: Rgba) -> Self {
+        Self(vec![(0f32, color_a), (1f32, color_b)])
     }
 
-    pub fn add_color_at_t(&mut self, t: f32, color: Rgb) {
-        let t = clamp_f32(t, 0.0, 1.0);
+    pub fn add_color_at_t(&mut self, t: f32, color: Rgba) {
+        let t = t.clamp(0.0, 1.0);
         self.0.push((t, color));
         self.0.sort_by(|a, b| {
             a.0.partial_cmp(&b.0)
                 .expect("Can't compare A and B because floats are tough")
         });
     }
 
-    pub fn color_at_t(&self, t: f32) -> Rgb {
-        let t = clamp_f32(t, 0.0, 1.0);
+    pub fn color_at_t(&self, t: f32) -> Rgba {
+        let t = t.clamp(0.0, 1.0);
         let (before_color, after_color) = self.get_bounding_colors_for_t(t);
         let mapped_t = map_t_of_range_a_to_range_b(t, before_color.0, after_color.0, 0.0, 1.0);
 
-        let [r1, g1, b1] = before_color.1;
-        let [r2, g2, b2] = after_color.1;
+        let [r1, g1, b1, ..] = before_color.1;
```

Comment:
is this being expanded to the a1 value or are there more entries in these?

Reply:
I was ignoring alpha because I didn't want to spend cycles interpolating between 1.0 and 1.0

I changed my tune and simplified the whole thing, let me know what you think.